r2dbc-core/src/main/java/com/google/cloud/sql/core/GcpConnectionFactoryProvider.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2020 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql.core;
+
+import static io.r2dbc.spi.ConnectionFactoryOptions.Builder;
+import static io.r2dbc.spi.ConnectionFactoryOptions.DRIVER;
+import static io.r2dbc.spi.ConnectionFactoryOptions.HOST;
+import static io.r2dbc.spi.ConnectionFactoryOptions.PORT;
+import static io.r2dbc.spi.ConnectionFactoryOptions.PROTOCOL;
+
+import io.netty.handler.ssl.SslContextBuilder;
+import io.r2dbc.spi.ConnectionFactory;
+import io.r2dbc.spi.ConnectionFactoryOptions;
+import io.r2dbc.spi.ConnectionFactoryProvider;
+import java.util.Properties;
+import java.util.function.Function;
+
+/** {@link ConnectionFactoryProvider} for proxied access to GCP Postgres and MySQL instances. */
+public abstract class GcpConnectionFactoryProvider implements ConnectionFactoryProvider {
+
+  abstract ConnectionFactory tcpConnectonFactory(
+      Builder optionBuilder, Function<SslContextBuilder, SslContextBuilder> customizer);
+
+  abstract ConnectionFactory socketConnectionFactory(Builder optionBuilder, String socket);
+
+  abstract Builder createBuilder(ConnectionFactoryOptions connectionFactoryOptions);
+
+  abstract boolean supportedProtocol(String protocol);
+
+  @Override
+  public ConnectionFactory create(ConnectionFactoryOptions connectionFactoryOptions) {
+    String connectionName = connectionFactoryOptions.getRequiredValue(HOST);
+    String protocol = connectionFactoryOptions.getRequiredValue(PROTOCOL);
+
+    Properties properties = new Properties();
+    properties.put(CoreSocketFactory.CLOUD_SQL_INSTANCE_PROPERTY, connectionName);
+
+    if (supportedProtocol(protocol)) {
+      return createFactory(connectionFactoryOptions, properties);
+    }
+
+    throw new UnsupportedOperationException(
+        "Cannot create ConnectionFactory " + connectionFactoryOptions);
+  }
+
+  ConnectionFactory createFactory(
+      ConnectionFactoryOptions connectionFactoryOptions, Properties properties) {
+    Builder optionBuilder = createBuilder(connectionFactoryOptions);
+
+    // precompute SSL Data to avoid blocking calls during the connect phase.
+    CoreSocketFactory.getSslData(properties);
+    String socket = CoreSocketFactory.getUnixSocketArg(properties);
+
+    if (socket != null) {
+      return socketConnectionFactory(optionBuilder, socket);
```

Comment:
Could you clarify your question?

Reply:
This function returns a ConnectionFactory, presumable that is used to create functions. Presumably, this means this factory is reused for multiple connections - maybe even all connections? If so, I'm not sure having two different ConnectionFactories makes sense, since the user may have one connection that uses the socket arg and one that doesn't. Should there be a single ConnectionFactory that has a connect method that looks at socket to determine if it should return a socket or a regular connection?

Reply:
Do you mean we should cache the socket factory and return it if it exists?